src/main/reseal.c

```diff
@@ -279,126 +282,141 @@ int main(int argc, char **argv)
-  size_t output_length = 0;
+  uint8_t *unseal_output = NULL;
+  size_t unseal_output_len = 0;
 
   int *pcrs = NULL;
   int pcrs_len = 0;
 
   if (parse_pcrs_string(pcrsString, &pcrs, &pcrs_len) != 0 || pcrs_len < 0)
   {
     kmyth_log(LOG_ERR, "failed to parse PCR string %s ... exiting", pcrsString);
-    free(outPath);
-    free(output);
+    free(unseal_output);
     free(pcrs);
     return 1;
   }
 
-  // Call top-level "kmyth-reseal" function
-  if (tpm2_kmyth_seal_file(inPath, &output, &output_length,
-                           (uint8_t *) authString, auth_string_len,
-                           (uint8_t *) ownerAuthPasswd, oa_passwd_len,
-                           pcrs, (size_t)pcrs_len, cipherString, expected_policy,
-                           bool_trial_only))
+ // Call top-level "kmyth-unseal" function
+  if (tpm2_kmyth_unseal_file(inPath, &unseal_output, &unseal_output_len,
+                             (uint8_t *) authString, auth_string_len,
+                             (uint8_t *) ownerAuthPasswd, oa_passwd_len,
+                             bool_policy_or))
   {
-    kmyth_log(LOG_ERR, "kmyth-reseal error ... exiting");
+    kmyth_clear_and_free(unseal_output, unseal_output_len);
+    kmyth_log(LOG_ERR, "kmyth-unseal failed ... exiting");
     kmyth_clear(authString, auth_string_len);
     kmyth_clear(ownerAuthPasswd, oa_passwd_len);
+    return 1;
+  }
+  
+  uint8_t *seal_output = NULL;
+  size_t seal_output_len = 0;
+
+  // Call top-level "kmyth-seal" function
+if (tpm2_kmyth_seal(unseal_output, unseal_output_len, &seal_output, &seal_output_len,
+                      (uint8_t *) authString,
+                      auth_string_len,
+                      (uint8_t *) ownerAuthPasswd,
+                      oa_passwd_len,
+                      pcrs,
+                      (size_t) pcrs_len,
+                      cipherString,
+                      expected_policy,
+                      bool_trial_only))
+  {
+    kmyth_log(LOG_ERR, "kmyth-seal error ... exiting");
+    kmyth_clear(authString, auth_string_len);
+    kmyth_clear(ownerAuthPasswd, oa_passwd_len);
+    kmyth_clear_and_free(unseal_output, unseal_output_len);
     free(pcrs);
-    free(outPath);
-    free(output);
     return 1;
   }
 
+  kmyth_clear_and_free(unseal_output, unseal_output_len);
   kmyth_clear(authString, auth_string_len);
   kmyth_clear(ownerAuthPasswd, oa_passwd_len);
 
-  if (write_bytes_to_file(outPath, output, output_length))
+   // rename input file to <input filename>.orig to preserve it
+  char * renamePath = malloc(strlen(inPath) + strlen(".orig") + 1);
+  strncpy(renamePath, inPath, strlen(inPath));
+  strncat(renamePath, ".orig", 5);
+  if (!stat(renamePath, &st) && !forceOverwrite)
   {
-    kmyth_log(LOG_ERR, "error writing data to .ski file ... exiting");
-    free(outPath);
-    free(output);
-    free(pcrs);
+    kmyth_log(LOG_ERR,
+              "output filename (%s) already exists ... exiting",
+              renamePath);
```

Comment:
I think we need to `free(seal_output)` and `free(renamePath)` before returning due to this error.